src/librbd/io/AioCompletion.cc

```diff
@@ -273,11 +272,8 @@ void AioCompletion::complete_event_socket() {
 }
 
 void AioCompletion::notify_callbacks_complete() {
-  state = AIO_STATE_COMPLETE;
-
-  {
-    std::unique_lock<std::mutex> locker(lock);
-    cond.notify_all();
+  if (waiting_for_complete.load(std::memory_order_acquire)) {
```

Comment:
Do we still need to bother with `waiting_for_complete` now that that there is no mutex around here?  Can you see if notifying `cond` unconditionally makes a difference in your measurements?

It would be nice to update the description of the PR and post a new set of flamegraphs since the approach has changed quite radically.  Of the three changes that are being talked about there, only one is up-to-date:
> 3. Relax the memory order of the atomic operations to update the AioCompletion state.

Reply:
The original reason for `waiting_for_complete` was to avoid the performance cost of taking the lock and notifying the conditional in the normal I/O path (e.g. when running `rbd bench`, as I did to take these measurements) where the conditional is not actually used.

Removing the lock partially alleviates that problem, but I'm guessing there is still a performance cost to using the atomic wait+notify when it's not needed? I can try it though and take some more measurements to confirm.

I did take some more measurements and flame graphs with the new approach and found that the performance gain was more or less the same for both, since either way `waiting_for_complete` prevents us from going down the conditional path when running `rbd bench` (so in the original approach we don't take the lock, in the new approach we don't do the atomic wait+notify).